### PR TITLE
feat(Core/Threading): replace ace threading

### LINKED ISSUE #99: Core/Threading: Rewrite Threading system [$150]
I open this issue to keep trace of the status for threading system rewriting

In few words we need to switch from ACE to c++11 libraries ( not boost ) 

cmangos do it 1 year ago with following main commits:
- [ ]  [thread rewriting](https://github.com/cmangos/mangos-wotlk/commit/b73488d94303fa0c23005f802e749300217fb398) 
- [x]  [atomic types](https://github.com/cmangos/mangos-wotlk/commit/835e2a02b11a5f4e511461b7ff790190b3f83de3)
- [ ]  [mutexes and locks](https://github.com/cmangos/mangos-wotlk/commit/8ae67274a754f821f0bea6a5f771eab3df764a57)

I hope someone can help
## 

<bountysource-plugin>

---
There is a **[$150 open bounty](https://www.bountysource.com/issues/37059511-core-threading-rewrite-threading-system?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Replace ACE threading

## Issues Addressed:
- Closes #99
- Part #3451
- After merge need apply https://github.com/ElunaLuaEngine/Eluna/pull/348

## Tests Performed:
- Windows `3 people`
image
- In Unix `2 people`

## How to Test the Changes:
- Check enter in game (priority hight)
- Uptime `> 5-10 min`

## Target Branch(es):
- [x] Master

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/common/Threading/LockedQueue.h [hunk 1]
```diff
@@ -7,137 +7,134 @@
...
-        {
-        }
+        unlock();
+    }
 
-        //! Destroy a LockedQueue.
-        virtual ~LockedQueue() = default;
+    //! Adds items back to front of the queue
+    template<class Iterator>
+    void readd(Iterator begin, Iterator end)
```
Comment (claudiodfc) [inline src/common/Threading/LockedQueue.h:50]: is this name written as **readd** on purpose?

**Comment by claudiodfc** [comment]
Apparently on Linux std::mutex works just fine but on Windows has some issues?

more about it here
https://stackoverflow.com/questions/14191566/c-mutex-in-namespace-std-does-not-name-a-type

File: src/common/Threading/LockedQueue.h (hunk 1 above)
Comment (Winfidonarleyan) [reply src/common/Threading/LockedQueue.h:50]: Copy from https://github.com/TrinityCore/TrinityCore/blob/b94851b824f6e4d7f7a7cd01a4c99e02b1e86b97/src/common/Threading/LockedQueue.h#L61

**Comment by Winfidonarleyan** [comment]
No problem in Win

**Comment by Mitradis** [comment]
ready?

**Comment by Winfidonarleyan** [comment]
In win maybe yes

**Comment by Mitradis** [comment]
Compiling and work, but crashes on ctrl+c still happened.

**Comment by Mitradis** [comment]
i help with modules. Ok we can merge this i think.

**Comment by Winfidonarleyan** [comment]
Okay, totaly replaced

**Comment by Winfidonarleyan** [comment]
Ready

**Comment by Mitradis** [comment]
this Pr should help with ctrlc crash? or this problem belongs to another part Ace remove?

Tested, work.

**Comment by Winfidonarleyan** [comment]
> this Pr should help with ctrlc crash? or this problem belongs to another part Ace remove?

No help with `CTRL+C`. Replace thread only

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @azerothcore/developers @mpfans @pklloveyou please check this PR

**Comment by claudiodfc** [comment]
## Tested on Windows 10
Overall no errors or weird behaviour was detected on a 13 minutes uptime server.

- Played a bit around character creation and adding a few items using commands for a class.
- Learnt spells and fought a Gorilla in Scholazar Basin
- Tamed Loque'naraq
- Uptime at the screenshot: 13 minutes

image

**Comment by FrancescoBorzi** [comment]
good job @Winfidonarleyan and thank you very much @claudiodfc and @Mitradis  for testing this.

As this is quite a big change, I'd like to have more people test this before merging.

**Comment by Winfidonarleyan** [comment]
Need test in `UNIX` system

**Comment by Winfidonarleyan** [comment]
Okay, go detailed test
# Ununtu 20.04
- Characters created - work
- Characters loading - work
- Teleport (command) - work
- Chat - work
- instance scripts (ICC) - work
- Mails - work
- Auction - work
- Character save - work
- Uptime - `> 20 min`

**Comment by Zoidwaffle** [comment]
Tested on Debian Testing as well.

Screenshot from 2021-03-17 14-51-05

Uptime 16 minutes.

Teleported all around Teldrassil and visited all herbal-spots. 23 herbs found :)

**Comment by claudiodfc** [comment]
```
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/GlobalMethods.h:204:53: fatal error: no member named 'LockType' in 'HashMapHolder<Player>'
            ACORE_READ_GUARD(HashMapHolder<Player>::LockType, *HashMapHolder<Player>::GetLock());
                             ~~~~~~~~~~~~~~~~~~~~~~~^
1 error generated.
```
I believe the mutex has to be done for the Lua engine aswell

**Comment by Zoidwaffle** [comment]
I have now seen the same issue twice. After some time in the game (not doing anything, I'm logged in as a GM), I lose contact to the server. So it acts like a disconnect - I'm inside the game, but unable to interact with anything and I can't log out either.

I don't recall ever seeing that before with AC, so I assume it's relevant.

From the worldserver:
`WorldSocket::handle_input_header(): client (account: 0, char [GUID: 0, name: <none>]) sent malformed packet (size: 18245, cmd: 539959380)`

And I can't exit the worldserver pressing ctrl+c, it just hangs.

There's nothing else in the logfiles - perhaps I need to increase verbosity or something... Or does this make some sort of sense as is?

**Comment by Zoidwaffle** [comment]
OK, some more info. It happened 4 times now.

It appears as if the error happens when my character should log out normally.  Then the server dies. I'm unable to rejoin the server  again as the same or another account.

Tried as a normal account, just to see if the GM part is important or not... And the same thing happened.

**Comment by FrancescoBorzi** [comment]
> OK, some more info. It happened 4 times now.
> 
> It appears as if the error happens when my character should log out normally. Then the server dies. I'm unable to rejoin the server again as the same or another account.
> 
> Tried as a normal account, just to see if the GM part is important or not... And the same thing happened.

@Zoidwaffle well done

@Winfidonarleyan please fix it and let us know when it's ready again

**Comment by Zoidwaffle** [comment]
If I can run it with some debugging or whatever enabled to help out, just let me know what to do.

**Comment by Winfidonarleyan** [comment]
@Zoidwaffle how to, please steps

**Comment by Zoidwaffle** [comment]
`AzerothCore rev. a03d7b2c8a8f 2021-03-18 20:56:45 +0700 (mutex branch) (Unix, Release) (worldserver-daemon) ready...`

After recompiling I left it for maybe 30 minutes. It didn't allow me to log in. So I restarted it.

Started 20:34:46. I logged in... didn't do much - took a look at the real estate prices in Orgrimmar. Decided it was too expensive and looked at some native cottages in the Sen'Jin suburbs instead. To be accepted I would have to get some tusk-implants... So a lot to consider.

At 20:54:46 the server died. Kinda suspicious at a 20 minute mark.

**Comment by Winfidonarleyan** [comment]
Okay, i try test today

**Comment by Winfidonarleyan** [comment]
Okay, i don't know how to repeat you bug.
I try uptime 30+ min in Win

image
image

--- 
Start test in Ubuntu 20.04

**Comment by Winfidonarleyan** [comment]
`Ubuntu 20.04`

image

---
So, i don't know how to repeat :/

**Comment by Zoidwaffle** [comment]
Hey, I will try to rebuild and test again. I compile using `-DENABLE_EXTRAS=1 -DENABLE_EXTRA_LOGS=1` but it doesn't seems to make much difference.

**Comment by Zoidwaffle** [comment]
```
AC>AzerothCore rev. b9afa3d94f10 2021-03-23 16:03:23 +0700 (mutex branch) (Unix, Release)
Connected players: 0. Characters in world: 0.
Connection peak: 0.
Server uptime: 6 second(s).
Update time diff: 10ms, average: 21ms.
```

It's up and running now... This time I can break it using "ctrl+c", so that's an improvement. I will check back in half an hour or so with an update!

**Comment by Winfidonarleyan** [comment]
Okay, i wait you

**Comment by Zoidwaffle** [comment]
Hm, it died after 12 minutes only. The issue with unable to quit using "ctrl+c" is back as well, seems like it's part of what happens when the server dies.

Tried again - uptime was 33 minutes this time.

I do use various modules:
```
mod-arac
mod-ah-bot
mod-autobalance
mod-bg-slaveryvalley
mod-transmog
mod-weapon-visual
mod-morphsummon
mod-congrats-on-level
mod-npc-buffer
mod-account-achievements
```

I'm running Debian Testing - I used to have Debian 10, but some of the recent changes to map-tools forced me into using Testing repositories. I can change to Ubuntu 20.04 if that's an "official" OS for AC.

But right now I will try out Debian 10 again, without any modules or other modifications.

**Comment by Winfidonarleyan** [comment]
@Zoidwaffle try clean core

**Comment by Zoidwaffle** [comment]
Sorry, I'm stuck at an error not relevant for this PR... But warden is failing for me now.

```
    /home/azerothcore/azerothcore-wotlk/acore_source/src/server/game/Warden/WardenWin.cpp:108:55: fatal error: too few arguments to function call, expected at least 2, have 1; did you mean '::ByteArrayToHexStr'?
        sLog->outDebug(LOG_FILTER_WARDEN, "C->S Key: %s", acore::Impl::ByteArrayToHexStr(_inputKey).c_str());
                                                          ^~~~~
                                                          ::ByteArrayToHexStr
    /home/azerothcore/azerothcore-wotlk/acore_source/src/common/Utilities/Util.h:356:13: note: '::ByteArrayToHexStr' declared here
    std::string ByteArrayToHexStr(Container const& c, bool reverse = false)
```

I might have time to look into it later in the evening, but if I'm the only one having issues then it's probably my server / OS and not  general thing.

**Comment by Winfidonarleyan** [comment]
in `build` dir

```shell
git reset --hard b9afa3d94f1043c19812887e9f2f846f85f5dd04
rm -rf *
````

**Comment by Winfidonarleyan** [comment]
@Zoidwaffle have news?

**Comment by Zoidwaffle** [comment]
@Winfidonarleyan : I took the master branch, commit 210530, it was stable for 14 hours, until I stopped it.

Then I recompiled this branch, AzerothCore rev. 52d02c0dff31 2021-03-24 13:19:08 +0700 (mutex branch) (Unix, Release), and it died in 14 minutes. However, it was with modules...

And once more, this time clean... no modules, no previous database-data, no nothing... I get some errors:

`ERROR: > Config: Missing name SetDeprecatedExternalPasswords in config, add "SetDeprecatedExternalPasswords = 0"` (it seems this should be fixed in a coming PR already).
`ERROR: Can't set worldserver process priority class, error: Permission denied` (is it trying to set a `nice` level?)

I can see both errors also existed in the build done using master, so they can probably be ignored.

It ended up dying after ~20-25 minutes.

Is the "After merge need apply ElunaLuaEngine/Eluna#348" needed? I'm not sure how to do that, those files don't seem to exist already?

**Comment by Zoidwaffle** [comment]
If I'm the only one with this issue, I can change to Ubuntu. I'm already forced into using Debian Testing anyway... (I personally prefer Debian and I would prefer to stay with a stable release... but I'm flexible as long as it remains Linux-based).

File: src/server/game/Globals/ObjectAccessor.cpp [hunk 2]
```diff
@@ -26,6 +26,35 @@
...
+void HashMapHolder<T>::Remove(T* o)
+{
+    std::unique_lock<std::shared_mutex> lock(*GetLock());
+    m_objectMap.erase(o->GetGUID());
+}
+
+template<class T>
+T* HashMapHolder<T>::Find(uint64 guid)
+{
+    std::unique_lock<std::shared_mutex> lock(*GetLock());
```
Comment (Rochet2) [inline src/server/game/Globals/ObjectAccessor.cpp:46]: This used to be shared_lock

**Comment by claudiodfc** [comment]
I tested this aswell and the same happens to me:

```
2021-03-25 12:31:24 Command: .server info [Player: Thor (3l) (Account: 1) X: 3046.301514 Y: -3395.162354 Z: 299.391083 Map: 533 (Naxxramas) Area: 3456 (Naxxramas) Zone: Unknown Selected:  (0l)]
2021-03-25 12:35:00 Average update time diff: 10. Players online: 0.
2021-03-25 12:45:24 Auction House Listing thread exiting without problems.
```

The server was running well for a solid 10 minutes and then I decided to log out and see how long it would last. At 12:45 tested again a `.server info` but It no longer responded. I hit CTRL+C and the Auction house listing thread exited. However, I had to hit CTRL+C again for the server to close.

**Comment by Zoidwaffle** [comment]
@claudiodfc : sounds like what I see. I can also exit the AH nicely and then it just hangs when crashed.

File: src/server/game/Globals/ObjectAccessor.cpp (hunk 2 above)
Comment (Winfidonarleyan) [reply src/server/game/Globals/ObjectAccessor.cpp:46]: Okay, thanks

**Comment by Zoidwaffle** [comment]
@Winfidonarleyan : I'm recompiling your latest version now :+1:

**Comment by Zoidwaffle** [comment]
`AzerothCore rev. c23fc113f999 2021-03-25 21:33:42 +0700 (mutex branch) (Unix, Release)` uptime just passed 40 minutes, which is a new record. I will update later in the evening if it keeps working or not.

**Comment by Zoidwaffle** [comment]
```
AC>AzerothCore rev. c23fc113f999 2021-03-25 21:33:42 +0700 (mutex branch) (Unix, Release)
Connected players: 0. Characters in world: 0.
Connection peak: 1.
Server uptime: 4 hour(s) 1 minute(s) 43 second(s).
```

Yep, it seems good. Well done! :rocket:

**Comment by Winfidonarleyan** [comment]
So, all okay?

**Comment by Zoidwaffle** [comment]
@Winfidonarleyan : yes, it seems to work just fine for me now. Thanks!

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan the build is failing

**Comment by Winfidonarleyan** [comment]
> @Winfidonarleyan the build is failing

ElunaLuaEngine/Eluna#348

**Review by Yehonal — APPROVED** [review]
GJ! Have Someone tested it on a live server with medium population?

**Comment by FrancescoBorzi** [comment]
waiting for @Footman @mpfans and @pklloveyou to test this with players

**Comment by mpfans** [comment]
I've been testing it for a week and I think we can merge it

**Comment by Winfidonarleyan** [comment]
@mpfans Thanks for testing

**Pull request merged**